### DISCUSSION TITLE
Forge does not always compute necessary recompilations correctly

### DISCUSSION THREAD

**Discussion by aathan** [foundry-rs/foundry#1299] [Help] [answered]
I've noticed intermittent cases where forge doesn't seem to be recompiling contracts I expect need recompilation. I've also seen this strange error:

```
$ forge test -vvv --fork-url http://127.0.0.1:8545/ -m testAllPublic
ignoring ``: invalid filter directive
```

**Comment by onbjerg** [comment]
What platform are you on and what version? Can you also provide an example we can test off of?

**Reply by aathan** [reply]
Not really, sinceit's happening while working on a large project I'm working on and I have no idea what's producing it and therefore no idea how to reproduce it independently of the entire project I'm working on.  I thought maybe the error might look familiar to you as something one of the packages you are using emits...

```
Linux ubuntu 5.13.0-35-generic #40-Ubuntu SMP Mon Mar 7 08:03:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ forge --version
ignoring ``: invalid filter directive
forge 0.2.0 (64087b5 2022-04-13T00:10:42.429318924+00:00)
```

Oh look at that, it happened with just `forge --version`  ... I know I did a forgeup recently.

**Reply by onbjerg** [reply]
That's weird, I can't reproduce with the latest nightly:

```
forge 0.2.0 (a5e58ad 2022-04-14T13:03:55.667821178+00:00)
```

The error is from the tracing crate

I'm on Linux too:

```
Linux bb 5.16.14-1-MANJARO #1 SMP PREEMPT Fri Mar 11 14:12:18 UTC 2022 x86_64 GNU/Linux
```

It's going to be really hard to figure out what the underlying bug for recompilations is though since we have no way to test.

**Reply by aathan** [reply]
Yes of course. That one, if I can reproduce a specific set of actions that causes certain things not to recompile, I'll add to this conversation.

**Reply by aathan** [reply]
So I'm hitting this again. This time I have a little more information. So, I'm doing:

```
forge test -vvv --fork-url http://127.0.0.1:8545/ -m testX
[⠢] Compiling...
[⠃] Compiling 1 files with 0.8.13
[⠒] Solc finished in 683.97ms
Compiler run successful
```
... notice the test does not run. A prior compilation emitted some warnings in the contract being testing (not Contract.t.sol) ...

```
forge test -vvv --fork-url http://127.0.0.1:8545/ -m testX
[⠒] Compiling...
No files changed, compilation skipped
```

Here I just "touched" the test file, but I guess the compilation is based on content delta, so I had to go in and add a space somewhere to force the recompilation. Once I did that, the next repeat of that same forge command above, ran the test.

I'm wondering if there's a bug where if there's a compilation error (or warning) in the core (non test) contracts but no error in the test contract, then you fix the core contract, somehow the test contract is in a state that doesn't run correctly and/or fails silently until you recompile the test.  E.g., maybe it is incorrectly being marked as not-dirty even though it was further down the compilation chain when the core error was detected, and then doesn't get recompiled until you also touch it.

... hmm, but I just tried that chain of events and that wasn't it. Scratching my head.

**Reply by aathan** [reply]
This may be my "user error" (my fault) but combined with the delta mechanism, it creates somewhat unexpected results ... the sequence of events could be explained by there being an error in the test contract AND the core contract. If you then fix the core contract, and fix the test contract but don't hit save, your next `forge test...` will recompile the core contract but *not* the test contract (because it has not changed).

The net result is that your forge test emits no test output, instead of having retried the compilation of the test contract. Follow?

Arguably, failed compilations should leave those contracts marked as dirty so they will be recompiled again on subsequent forge commands. If you're trying to minimize emitting the same errors repeatedly, maybe there can be two flags `-f` to force recompile everything and `-F` to force recompilation on contracts previously not compiled due to errors.

**Reply by onbjerg** [reply]
Can you add this context to https://github.com/foundry-rs/foundry/issues/1244 and possibly a separate issue about the compiler run successful -> no test are run (unless another issue is already present, then add context to that)? Discussions are primarily to discuss new features/user help and it seems like you might have found an actual bug

**Reply by aathan** [reply]
Sure. I didn't want to do the converse and discuss non-bugs in a bug report. If you think this is a bug I'll open a bug/add to one.

**Reply by onbjerg** [reply]
I think it is - going to mark a dummy answer as the solution so we can continue the discussion in the issue tracker instead :)

**Comment by aathan** [comment]
Hey, so:

```
$ export RUST_LOG=
$ forge --version
ignoring ``: invalid filter directive
forge 0.2.0 (64087b5 2022-04-13T00:10:42.429318924+00:00)
```

Empty RUST_LOG is the problem.

**Comment by onbjerg** [comment, marked as answer]
After discussion in https://github.com/foundry-rs/foundry/discussions/1299#discussioncomment-2568880 this seems like a bug, possibly related to https://github.com/foundry-rs/foundry/issues/1244
